### PR TITLE
Update to 24.1.2, use pip source to install itself.

### PR DESCRIPTION
**Pull request by jjhelmus**
Checklist
* [x] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [ ] Bumped the build number (if the version is unchanged)
* [x] Reset the build number to `0` (if the version changed)
* [x] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [x] Ensured the license file is being packaged.

Changes:
* Update to version 24.1.2
* Bump minimum Python version to 3.8. Python 3.7 support was removed in version 24.1.
* Use the pip source to install itself. Note that setup.py was removed in version 24.1.
* Remove the run requirements on `setuptools` and `wheel`. They are only needed to build the package not at runtime.

### CODE REVIEW

**Comment by jjhelmus** [comment]
@conda-forge-admin, please rerender

**Comment by jjhelmus** [comment]
Another option for building pip is to use the wheel published on PyPI. This would remove the host requirement on `setuptools` and `wheel` which breaks a dependency cycle. 

I have a [branch](https://github.com/jjhelmus/pip-feedstock/tree/cf_v24.1.2_whl) on my fork that implements this if there is interest.

File: recipe/meta.yaml [hunk 1]
```diff
@@ -1,12 +1,12 @@
...
 
 package:
   name: pip
   version: {{ version }}
 
 source:
-  url: https://pypi.io/packages/source/p/pip/pip-{{ version }}.tar.gz
-  sha256: ea9bd1a847e8c5774a5777bb398c19e80bcd4e2aa16a4b301b718fe6f593aba2
+    url: https://pypi.io/packages/source/p/pip/pip-{{ version }}.tar.gz
+    sha256: e5458a0b89f2755e0ee8c0c77613fe5273e05f337907874d64f13171a898a7ff
```
Comment (jakirkham) [inline recipe/meta.yaml:9]: nit: Let's keep the same indentation 

```suggestion
  url: https://pypi.io/packages/source/p/pip/pip-{{ version }}.tar.gz
  sha256: e5458a0b89f2755e0ee8c0c77613fe5273e05f337907874d64f13171a898a7ff
```

**Review by jakirkham — COMMENTED** [review]
Thanks Jonathan! 🙏

Also nice to see you again 😄

Think this approach is great and we should do it

Had a very minor suggestion below

File: recipe/meta.yaml (hunk 1 above)
Comment (jjhelmus) [reply recipe/meta.yaml]: Done.

File: recipe/build.sh [hunk 2]
```diff
@@ -1,6 +1,7 @@
 #!/bin/bash
 
-$PYTHON setup.py install --single-version-externally-managed --record record.txt
+# use the pip source to install itself
+PYTHONPATH="./src" $PYTHON -m pip install . --no-deps --no-build-isolation -vv
```
Comment (jakirkham) [inline recipe/build.sh:4]: These flags are [already set for us by `conda-build`]( https://github.com/conda/conda-build/blob/417d3f4c73ba525c037c1f453da24df8b1ee12aa/conda_build/build.py#L3052-L3072 )

```suggestion
PYTHONPATH="./src" $PYTHON -m pip install . -vv
```

File: recipe/bld.bat [hunk 3]
```diff
@@ -1,4 +1,5 @@
-%PYTHON% setup.py install --single-version-externally-managed --record record.txt
+set PYTHONPATH=".\src"
+%PYTHON% -m pip install . --no-deps --no-build-isolation -vv
```
Comment (jakirkham) [inline recipe/bld.bat:2]: Same story on Windows (though [different part of the conda-build code base]( https://github.com/conda/conda-build/blob/417d3f4c73ba525c037c1f453da24df8b1ee12aa/conda_build/windows.py#L296-L317 ))

```suggestion
%PYTHON% -m pip install . -vv
```

**Review by jakirkham — COMMENTED** [review]
Thanks Jonathan! 🙏

Tidying a couple more things below

**Comment by beckermr** [comment]
@conda-forge/pip @conda-forge/core The removal of setuptools and wheel is breaking expectations people have had for what they get in the env when they install pip. Do we need to add them back?

**Comment by jakirkham** [comment]
The change was intentional in the bullet points above:

> * Remove the run requirements on `setuptools` and `wheel`. They are only needed to build the package not at runtime.

That said, no strong feelings if folks would like to add them back

**Comment by beckermr** [comment]
Yep. I see that the change was intentional. 

I am looking for further discussion. Do we want to take the pure path here or go more "batteries included."

AFAIK, we have shipped setuptools+wheel as run deps with pip for ~8 years now (https://github.com/conda-forge/pip-feedstock/commit/5f1aff626e8714969fcb83e5431ab07639b47ff0) since this feedstock was created. 

Given how low this package sits in the typical dev stack, the removal of these deps is definitely non-trivial.

**Comment by jakirkham** [comment]
Could we please move this to the open issue?

**Pull request merged**